grpc-circuitbreaker-utils/src/main/java/org/hypertrace/circuitbreaker/grpcutils/resilience/ResilienceCircuitBreakerProvider.java

```diff
@@ -0,0 +1,73 @@
+package org.hypertrace.circuitbreaker.grpcutils.resilience;
+
+import io.github.resilience4j.circuitbreaker.CircuitBreaker;
+import io.github.resilience4j.circuitbreaker.CircuitBreakerConfig;
+import io.github.resilience4j.circuitbreaker.CircuitBreakerRegistry;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+import lombok.extern.slf4j.Slf4j;
+import org.hypertrace.circuitbreaker.grpcutils.CircuitBreakerConfigParser;
+
+/** Utility class to provide Resilience4j CircuitBreaker */
+@Slf4j
+class ResilienceCircuitBreakerProvider {
+
+  private final CircuitBreakerRegistry circuitBreakerRegistry;
+  private final Map<String, CircuitBreakerConfig> circuitBreakerConfigMap;
+  private final Map<String, CircuitBreaker> circuitBreakerCache = new ConcurrentHashMap<>();
+  private final List<String> disabledKeys;
+
+  public ResilienceCircuitBreakerProvider(
+      CircuitBreakerRegistry circuitBreakerRegistry,
+      Map<String, CircuitBreakerConfig> circuitBreakerConfigMap,
+      List<String> disabledKeys) {
+    this.circuitBreakerRegistry = circuitBreakerRegistry;
+    this.circuitBreakerConfigMap = circuitBreakerConfigMap;
+    this.disabledKeys = disabledKeys;
+  }
+
+  public Optional<CircuitBreaker> getCircuitBreaker(String circuitBreakerKey) {
+    if (disabledKeys.contains(circuitBreakerKey)
+        || !circuitBreakerConfigMap.containsKey(circuitBreakerKey)) {
+      return Optional.empty();
+    }
+    return Optional.of(
+        circuitBreakerCache.computeIfAbsent(
+            circuitBreakerKey,
+            key -> {
+              CircuitBreaker circuitBreaker = getCircuitBreakerFromConfigMap(circuitBreakerKey);
+              circuitBreaker
+                  .getEventPublisher()
+                  .onStateTransition(
+                      event ->
+                          log.info(
+                              "State transition: {} for circuit breaker {}",
+                              event.getStateTransition(),
+                              event.getCircuitBreakerName()))
+                  .onCallNotPermitted(
+                      event ->
+                          log.debug(
+                              "Call not permitted: Circuit is OPEN for circuit breaker {}",
+                              event.getCircuitBreakerName()))
+                  .onEvent(
+                      event ->
+                          log.debug(
+                              "Circuit breaker event type {} for circuit breaker name {}",
+                              event.getEventType(),
+                              event.getCircuitBreakerName()));
+              return circuitBreaker;
+            }));
+  }
+
+  public Optional<CircuitBreaker> getDefaultCircuitBreaker() {
+    return getCircuitBreaker(CircuitBreakerConfigParser.DEFAULT_THRESHOLDS);
```

Comment:
So I see one part of the confusion here. You're using this special key for the "shared" circuit breaker that's used when there is no key. We can use a special key for that, but it should be local to this provider - it's indepdendent of the config.

I think part of the confusion is that the term default is being overloaded a bit here. There's two separate things:
- the config to use if no key matches (which can be disabled - if no specific key config, don't circuit break)
- the shared circuit breaker to use if there is no key (which would be an instance of the former case - the config to use if no key matches).

Reply:
Thanks for your patience, i hope i have addressed all the cases now. Due to multiple scenarios, i feel implementation is slightly complicated, not sure if we use these different scenarios.